File: content/8-mcp/1-mcp/pydantic/2-client.py
# MCP client — Pydantic AI with first-class MCP support.
#
# Pydantic AI has first-class MCP support. The agent discovers tools when
# the MCP server subprocess starts. No manual tool conversion, no explicit
# session management.
#
# The client starts the MCP server as a subprocess (stdio transport).

import asyncio
from pydantic_ai import Agent
from pydantic_ai.mcp import MCPServerStdio

LLM_MODEL = "openai:gpt-5.4"

# MCPServerStdio wraps a subprocess. Pass it to toolsets — the agent
# discovers available tools at runtime via the MCP protocol.
# Compare with the Tools section where tools were @agent.tool functions.

server = MCPServerStdio("uv", args=["run", "server.py"])

agent = Agent(LLM_MODEL, toolsets=[server])

async def main():
    async with agent:
        result = await agent.run(
            "Who's the engineering manager and what's their email?",
        )
        print(result.output)

asyncio.run(main())
# -> list_employees("engineering")   → Alice Chen (E001), Bob Smith, Carol Wu
# -> get_employee("E001")            → alice@company.com
# "The engineering manager is Alice Chen. Her email is alice@company.com."

File: content/8-mcp/1-mcp/mastra/2-client.ts
// MCP client — Mastra with @mastra/mcp.
//
// Mastra has full MCP support — both server (see server.ts) and client.
// MCPClient manages the subprocess lifecycle; listTools() returns
// tools ready for any Mastra agent.
//
// The client starts the MCP server as a subprocess (stdio transport).

import { Agent } from "@mastra/core/agent";
import { MCPClient } from "@mastra/mcp";

const LLM_MODEL = "openai/gpt-5.4";

// MCPClient manages one or more MCP server connections. Each server entry
// specifies how to start the subprocess. listTools() discovers all tools
// across all configured servers.
// Compare with the Tools section where tools were createTool() objects.

const mcp = new MCPClient({
  id: "mcp-client",
  servers: {
    directory: {
      command: "bun",
      args: ["run", "server.ts"],
    },
  },
});

const tools = await mcp.listTools();

const agent = new Agent({
  name: "mcp-agent",
  instructions: "You are a helpful assistant.",
  model: LLM_MODEL,
  tools,
});

const result = await agent.generate(
  "Who's the engineering manager and what's their email?",
  { maxSteps: 5 },
);

console.log(result.text);
await mcp.disconnect();
// -> list_employees("engineering")   → Alice Chen (E001), Bob Smith, Carol Wu
// -> get_employee("E001")            → alice@company.com
// "The engineering manager is Alice Chen. Her email is alice@company.com."
